src/Utilities/ToolLocationHelper.cs

```diff
@@ -1381,6 +1386,133 @@ public static string GetLatestSDKTargetPlatformVersion(string sdkIdentifier, str
             }
         }
 
+        /// <summary>
+        ///  Returns folders in VS installs of specifid range of versions, starting with 15.0.
+        /// </summary>
+        /// <param name="minVersion">Optional. If specified, only VS instances with this or bigger version will be included.</param>
+        /// <param name="maxVersion">Optional. If specified, only VS instances with smaller versions will be included. For instance, 16.0 means that only 15.* versions will be included.</param>
+        /// <param name="subFolder">Optional. If specified, the returned list will contain [VSInstallDir][subFolder] paths</param>
+        /// <returns>Returns folders in VS installs in VS version descending order</returns>
+        public static IEnumerable<string> GetFoldersInVSInstalls(Version minVersion = null, Version maxVersion = null, string subFolder = null)
+        {
+            string versionRange = $"{minVersion?.ToString() ?? string.Empty};{maxVersion?.ToString() ?? string.Empty}";
+            string[] vsInstallFolders;
+
+            lock (s_locker)
```

Comment:
The locking pattern looks very amenable to ConcurrentDictionary.GetOrAdd

Reply:
Is it required? The other methods in this file use locker.

Reply:
I think it would be something to keep in mind if we start seeing contention on the lock, since the lock protects unrelated pieces of data.